src/skin/qml/qmlplayerproxy.cpp

```diff
@@ -0,0 +1,125 @@
+#include "skin/qml/qmlplayerproxy.h"
+
+#include "mixer/basetrackplayer.h"
+
+// clazy: excludeall=incorrect-emit
+
+#define PROPERTY_IMPL(TYPE, NAME, GETTER, SETTER)    \
+    TYPE QmlPlayerProxy::GETTER() const {            \
+        const TrackPointer pTrack = m_pCurrentTrack; \
+        if (pTrack == nullptr) {                     \
+            return TYPE();                           \
+        }                                            \
+        return pTrack->GETTER();                     \
+    }                                                \
+                                                     \
+    void QmlPlayerProxy::SETTER(const TYPE& value) { \
+        const TrackPointer pTrack = m_pCurrentTrack; \
+        if (pTrack != nullptr) {                     \
+            pTrack->SETTER(value);                   \
+            emit NAME##Changed();                    \
+        }                                            \
+    }
+
+namespace mixxx {
+namespace skin {
+namespace qml {
+
+QmlPlayerProxy::QmlPlayerProxy(BaseTrackPlayer* pTrackPlayer, QObject* parent = nullptr)
+        : QObject(parent), m_pTrackPlayer(pTrackPlayer) {
+    connect(m_pTrackPlayer,
+            &BaseTrackPlayer::loadingTrack,
+            this,
+            &QmlPlayerProxy::slotLoadingTrack);
+    connect(m_pTrackPlayer,
+            &BaseTrackPlayer::newTrackLoaded,
+            this,
+            &QmlPlayerProxy::slotTrackLoaded);
+    connect(m_pTrackPlayer,
+            &BaseTrackPlayer::playerEmpty,
+            this,
+            &QmlPlayerProxy::trackUnloaded);
+    connect(this, &QmlPlayerProxy::trackChanged, this, &QmlPlayerProxy::slotTrackChanged);
+}
+
+void QmlPlayerProxy::slotTrackLoaded(TrackPointer pTrack) {
+    m_pCurrentTrack = pTrack;
+    if (pTrack != nullptr) {
+        connect(pTrack.get(),
+                &Track::changed,
+                this,
+                &QmlPlayerProxy::trackChanged);
+        connect(pTrack.get(),
+                &Track::keyUpdated,
+                this,
+                &QmlPlayerProxy::keyChanged);
+        connect(pTrack.get(),
+                &Track::colorUpdated,
+                this,
+                &QmlPlayerProxy::colorChanged);
+    }
+    emit trackChanged();
+    emit trackLoaded();
+}
+
+void QmlPlayerProxy::slotLoadingTrack(TrackPointer pNewTrack, TrackPointer pOldTrack) {
+    Q_UNUSED(pNewTrack);
+    Q_UNUSED(pOldTrack);
+    const TrackPointer pTrack = m_pCurrentTrack;
+    if (pTrack != nullptr) {
```

Comment:
You can omit the noisy `!=/== nullptr` comparisons, unless you explicitly need to return a boolean value.

Reply:
Didn't leaving it out cause conversion warnings? Also, it's more explicit IMHO.

Reply:
This topic has a history. 
* original we don't care 
* @rryan has started to make it explicit
* @uklotzde has removed the explicit comparisons

I still don't care, but prefer the implicit bool conversion for my code, because it is part of the std pointer api: 
https://en.cppreference.com/w/cpp/memory/shared_ptr/operator_bool

I would prefer to go back to the "Don't care" state for more developer freedom .. ;-)

Reply:
The implicit version worked for both Qt and std managed pointers. That was the main reason for the preference. The explicit comparison with nullptr does not work for QSharedPointer and QScopedPointer afair.

Reply:
Is there any reason to use the Qt versions? I thought they were obsolete with C++ smart pointers.

Reply:
No mass refactorings, that's the reason why.

Reply:
The only reason for QShared and QWeak pointers is that they integrates with the Qt object tree that is available anyway When using QObjects.

Reply:
Apart from a built-in qobjectCast there is no documentation of any special behavior for QObject? Don't confuse it with QPointer which is different.